attest/activation.go

```diff
@@ -0,0 +1,273 @@
+	}
+
+	// Make sure the AIK has sane key parameters (Attestation can be faked if an AIK
+	// can be used for arbitrary signatures).
+	// We verify the following:
+	// - Key is TPM backed.
+	// - Key is TPM generated.
+	// - Key is a restricted key (means it cannot do arbitrary signing/decrypt ops).
+	// - Key cannot be duplicated.
+	// - Key was generated by a call to TPM_Create*.
+	if att.Magic != tpm20GeneratedMagic {
+		return errors.New("creation attestation was not produced by a TPM")
+	}
+	if (pub.Attributes & tpm2.FlagFixedTPM) == 0 {
+		return errors.New("AIK is exportable")
+	}
+	if ((pub.Attributes & tpm2.FlagRestricted) == 0) || ((pub.Attributes & tpm2.FlagFixedParent) == 0) || ((pub.Attributes & tpm2.FlagSensitiveDataOrigin) == 0) {
+		return errors.New("provided key is not limited to attestation")
+	}
+
+	// Verify the attested creation name matches what is computed from
+	// the public key.
+	match, err := att.AttestedCreationInfo.Name.MatchesPublic(pub)
+	if err != nil {
+		return err
+	}
+	if !match {
+		return errors.New("creation attestation refers to a different key")
+	}
+
+	// Check the signature over the attestation data verifies correctly.
+	pk := rsa.PublicKey{E: int(pub.RSAParameters.Exponent), N: pub.RSAParameters.Modulus}
+	signHashConstructor, err := pub.RSAParameters.Sign.Hash.HashConstructor()
+	if err != nil {
+		return err
+	}
+	hsh := signHashConstructor()
+	hsh.Write(p.AIK.CreateAttestation)
+	verifyHash, err := cryptoHash(pub.RSAParameters.Sign.Hash)
+	if err != nil {
+		return err
+	}
+
+	if len(p.AIK.CreateSignature) < 8 {
+		return fmt.Errorf("signature invalid: length of %d is shorter than 8", len(p.AIK.CreateSignature))
+	}
+
+	//TODO(jsonp): Decode to tpm2.Signature & use that, once PR to expose DecodeSignature() is in.
+	if err := rsa.VerifyPKCS1v15(&pk, verifyHash, hsh.Sum(nil), p.AIK.CreateSignature[6:]); err != nil {
+		return fmt.Errorf("could not verify attestation: %v", err)
+	}
+
+	return nil
+}
+
+// Generate returns a credential activation challenge, which can be provided
+// to the TPM to verify the AIK parameters given are authentic & the AIK
+// is present on the same TPM as the EK.
+func (p *ActivationParameters) Generate() (secret []byte, ec *EncryptedCredential, err error) {
+	if err := p.checkAIKParameters(); err != nil {
+		return nil, nil, err
+	}
+
+	if p.EK == nil {
+		return nil, nil, errors.New("no EK provided")
+	}
+
+	rnd, secret := p.Rand, make([]byte, activationSecretLen)
+	if rnd == nil {
+		rnd = rand.Reader
+	}
+	if _, err = io.ReadFull(rnd, secret); err != nil {
+		return nil, nil, fmt.Errorf("error generating activation secret: %v", err)
+	}
+
+	switch p.TPMVersion {
+	case TPMVersion12:
+		ec, err := p.generateChallengeTPM12(secret)
+		if err != nil {
+			return nil, nil, err
+		}
+		return secret, ec, nil
+	case TPMVersion20:
+		ec, err := p.generateChallengeTPM20(secret)
+		if err != nil {
+			return nil, nil, err
+		}
+		return secret, ec, nil
+	}
```

Comment:
need a default case and to error. Maybe:

```
var ec *EncryptedCredential
switch p.TPMVersion {
case TPMVersion12:
    ec, err = p.generateChallengeTPM12(secret)
case TPMVersion20:
    ec, err = p.generateChallengeTPM20(secret)
default:
    return nil, nil, fmt.Errorf("unrecognized TPM version")
}
if err != nil {
   return nil, nil, fmt.Errorf("generating challenge: %v", err)
}
return secret, ec, nil
```

Reply:
Done.